src/System.Text.ValueBuilder/System/Variant.cs

```diff
@@ -0,0 +1,444 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace System
+{
+    /// <summary>
+    /// <see cref="Variant"/> is a wrapper that avoids boxing common value types.
+    /// </summary>
+    public readonly struct Variant
```

Comment:
Alternative way so avoid the allocation for this and many other cases (e.g. delegates) would be introduce ability to annotate arguments as non-escaping. The JIT would be then able to reliably convert the heap allocation into stack allocation; and we would not need to introduce a duplicate types and APIs every time we need to get something predictably allocated on the stack.

We may need similar annotations to allow generic types instantiated over ref types.

Reply:
+ @AndyAyersMS  please include other JIT devs for visibility on the issue of stack-bound lifetime of GC objects.  I know we have had an intern project on this, but I believe that was just inference based.   I suspect Jan's suggestion is about a user declaration that effectively asks the type system to inure that the objects's lifetime is not allow to escape the current method, but the goal is the same (you can stack allocate it).   

I am assuming however, that we are sticking with the value-type approach here however, as getting the 'auto-stack-allocation' working is probably non-trivial.

Reply:
> Alternative way to avoid the allocation for this and many other cases (e.g. delegates) would be introduce ability to annotate arguments as non-escaping. 

That's essentially asking to introduce borrowing into the system. Went down that rabbit hole in Midori. Lots of dragons down there. For example: once you annotate parameters as non-escaping you can't call instance methods on them unless `this` is also marked as non-escaping. Doable but the annotation gets viral very fast. 

Let me see if I can dig up the write up I had on why this gets a bit out of control fairly fast.

Reply:
> We may need similar annotations to allow generic types instantiated over ref types.

This is doable. We didn't go this far in the original implementation because it was really hard to see a concrete need. If we can make a case for it I think it's pretty straight forward. The syntax is *hopefully* not in debate: `where T : ref struct`.

Reply:
For objects with no GC fields, allowing cross-method stack allocation (allocate on caller's stack, pass stack address to callees) is conceptually simple. But getting strong correctness guarantees from up-front annotations seems tricky. And we have to worry about things like profiler attach rewrites of callees that might not preserve non-escaping behavior. 

For objects with GC fields, we have to be wary of broad non-pay-for-play cost increases as writing to GC fields of the stack allocated object in any method (other than perhaps the allocating method) will now require checked barriers.  And if we don't limit which objects can be stack allocated then the cost spreads even more broadly as the jit must use checked barriers almost everywhere.

The current plan for escape analysis requires that all object touches end up in a single method via inlining. That doesn't seem tractable in this case.

cc @erozenfeld

Reply:
> And we have to worry about things like profiler attach rewrites of callees that might not preserve non-escaping behavior.

If we were to make the nonescaping annotations part of the platform, you do not need worry about random IL rewrites screwing things up anymore. The IL rewriters would have to learn to respect them. The same way they have to respect ref-like annotations that we have introduced some time ago - they cannot box ref-like structs, even though all structs were boxable before.

> if we don't limit which objects can be stack allocated

The point of annotations would be to get the guess work out of this and make what gets stack allocated both predictable and limited.

Reply:
It would be great if we could find a way to have annotations for non-escaping parameters.

Possible complications:
1. If we have only parameter annotations and no general borrowing concept (which is problematic for the reasons @jaredpar mentioned), the C# compiler will have to do local escape analysis and it's unlikely to be perfect (i.e., in some cases it won't be able to prove that parameters are non-escaping when in fact they are). It will be hard to specify the rules on when the annotations are legal.

2. In some cases the annotations won't help. For example, if an object is allocated and then the reference is assigned to a field of another object and that object is passed to a call, it won't be enough to know that the call parameter is not escaping and there won't be a way to specify that the parameter's field is not escaping.

Reply:
Would we consider having annotations that are partially or completely unvalidated? i.e. the developer promises the object won't escape, and if it does, things break, ala Unsafe.*.

Reply:
Escaping analysis isn't really sufficient. To avoid awkward performance problems, the annotation needs to be doesn't escape, and object pointed at is never mutated. Looking at the use cases here, it's a vaguely reasonable assumption, but its a hard one to be sure about, especially in the presence of something like profiler rejit. (As the mutation would trigger a need to use a checked write barrier instead of a normal write barrier, and we'd really rather not do that to stores to the heap.)

Reply:
> To avoid awkward performance problems, the annotation needs to be doesn't escape, and object pointed at is never mutated

Can we document that mutation of objects pointed to by non-escaping references is slower?

We have the same issue with struct today. It is super rare somebody notices that mutation of object references in structs is slower, and I do not think it is even documented anywhere.

Reply:
Mmm, possibly. In any case, if we want to go the route of annotations we need to implement the annotation before we make ISpanFormattable a public api, as the annotation needs to be on ISpanFormattable the this pointer used to invoke an instance of ISpanFormattable.

Reply:
> we need to implement the annotation before we make ISpanFormattable a public api

Agree.

In any case, it would be useful to have the overall no-allocation high-performance formatting story figured out before we start making parts of it public. `ISpanFormattable` worked ok as internal implementation detail, but it may not be the right type to set in stone to support the no-allocation high-performance string formatting story.